zcash_client_backend/src/fees/common.rs

```diff
@@ -353,35 +424,96 @@ where
+                    )
+                } else {
+                    (fee_with_change, NonNegativeAmount::ZERO)
+                };
+
             let simple_case = || {
                 (
-                    vec![ChangeValue::shielded(
-                        change_pool,
-                        proposed_change,
-                        change_memo.cloned(),
-                    )],
+                    (0usize..split_count.into())
+                        .map(|i| {
+                            ChangeValue::shielded(
+                                change_pool,
+                                if i == 0 {
+                                    // Add any remainder to the first output only
+                                    (*per_output_change.quotient()
+                                        + *per_output_change.remainder()
+                                        + excess_fee)
+                                        .unwrap()
+                                } else {
+                                    // For any other output, the change value will just be the
+                                    // quotient.
+                                    *per_output_change.quotient()
+                                },
+                                change_memo.cloned(),
+                            )
+                        })
+                        .collect(),
                     fee_with_change,
                 )
             };
 
-            let change_dust_threshold = dust_output_policy
+            let change_dust_threshold = cfg
+                .dust_output_policy
                 .dust_threshold()
-                .unwrap_or(default_dust_threshold);
+                .unwrap_or(cfg.default_dust_threshold);
 
-            if proposed_change < change_dust_threshold {
-                match dust_output_policy.action() {
+            if per_output_change.quotient() < &change_dust_threshold {
+                match cfg.dust_output_policy.action() {
                     DustAction::Reject => {
                         // Always allow zero-valued change even for the `Reject` policy:
                         // * it should be allowed in order to record change memos and to improve
                         //   indistinguishability;
                         // * this case occurs in practice when sending all funds from an account;
                         // * zero-valued notes do not require witness tracking;
                         // * the effect on trial decryption overhead is small.
-                        if proposed_change.is_zero() {
+                        if per_output_change.quotient().is_zero() {
                             simple_case()
                         } else {
-                            let shortfall =
-                                (change_dust_threshold - proposed_change).ok_or_else(underflow)?;
+                            let shortfall = (change_dust_threshold - *per_output_change.quotient())
```

Comment:
This change also doesn't look right to me. It should be the overall shortfall.

Reply:
Resolved in https://github.com/zcash/librustzcash/pull/1590/commits/d3b82ee51f7da5fdd3a0cb08cc81d328d58da0e3